xdsl/ir/core.py

```diff
@@ -914,18 +914,18 @@ class OpTraits(Iterable[OpTrait]):
     at the first use.
     """
 
-    _traits: frozenset[OpTrait] | Callable[[], tuple[OpTrait, ...]]
+    gen_traits: Callable[[], tuple[OpTrait, ...]]
+    _traits: frozenset[OpTrait] | None
 
-    def __init__(
-        self, traits: frozenset[OpTrait] | Callable[[], tuple[OpTrait, ...]]
-    ) -> None:
-        self._traits = traits
+    def __init__(self, gen_traits: Callable[[], tuple[OpTrait, ...]]) -> None:
+        self.gen_traits = gen_traits
+        self._traits = None
 
     @property
     def traits(self) -> frozenset[OpTrait]:
         """Returns a copy of this instance's traits."""
-        if callable(self._traits):
```

Comment:
I've been meaning to get rid of this for a while, it's a much more expensive check than the new mechanism